File: android/app/src/main/kotlin/com/onlystack/starterapp/core/sync/SyncQueue.kt
package com.onlystack.starterapp.core.sync

import java.util.Date
import kotlin.math.min
import kotlin.math.pow

data class SyncQueueItem(
    val change: SyncChange,
    val retryCount: Int = 0,
    val nextRetryAt: Date = Date(),
) {
    val isEligible: Boolean get() = retryCount < 5 && nextRetryAt <= Date()

    fun scheduleRetry(): SyncQueueItem {
        val delayMs = min(2.0.pow(retryCount.toDouble()) * 2_000, 300_000.0).toLong()
        return copy(
            retryCount = retryCount + 1,
            nextRetryAt = Date(System.currentTimeMillis() + delayMs),
        )
    }
}

class SyncQueue {
    private val items = mutableListOf<SyncQueueItem>()

    fun enqueue(change: SyncChange) { items.add(SyncQueueItem(change)) }

    fun dequeueEligible(): List<SyncQueueItem> = items.filter { it.isEligible }

    fun markFailed(ids: Set<String>) {
        val updated = items.map { if (it.change.id in ids) it.scheduleRetry() else it }
        items.clear()
        items.addAll(updated.filter { it.retryCount < 5 })
    }

    fun remove(ids: Set<String>) { items.removeAll { it.change.id in ids } }

    val isEmpty: Boolean get() = items.isEmpty()
}

File: ios/StarterApp/Core/Sync/SyncQueue.swift
import Foundation

struct SyncQueueItem {
    let change: SyncChange
    var retryCount: Int = 0
    var nextRetryAt: Date = .now

    // Exponential backoff: 2^n * 2 seconds, capped at 300s
    mutating func scheduleRetry() {
        retryCount += 1
        let delay = min(pow(2.0, Double(retryCount)) * 2.0, 300.0)
        nextRetryAt = Date.now.addingTimeInterval(delay)
    }

    var isEligible: Bool { retryCount < 5 && nextRetryAt <= .now }
}

final class SyncQueue {
    private var items: [SyncQueueItem] = []

    func enqueue(_ change: SyncChange) {
        items.append(SyncQueueItem(change: change))
    }

    func dequeueEligible() -> [SyncQueueItem] {
        items.filter(\.isEligible)
    }

    func markFailed(ids: Set<String>) {
        for i in items.indices where ids.contains(items[i].change.id) {
            items[i].scheduleRetry()
        }
        // Drop permanently failed items (>= 5 retries)
        items.removeAll { $0.retryCount >= 5 }
    }

    func remove(ids: Set<String>) {
        items.removeAll { ids.contains($0.change.id) }
    }

    var isEmpty: Bool { items.isEmpty }
}
